File: sortingAlgo/cocktailSort/cocktailSort.py
# cocktail sort function
def cocktailSort(A):
    up = range(len(A)-1)
    while True:
        for indices in (up, reversed(up)):
            
            # reseting the swapped flag
            swapped = False
            
            # looping through left to right
            for i in indices:
                if A[i] > A[i+1]:
                    A[i], A[i+1] =  A[i+1], A[i]
                    swapped = True
            
            # if noting is swapped means, array is sorted
            if not swapped:
                return
 
# initializing the array
array = [25,1,-2,70,100,-10,40,2,0,5]
print("Array before sorting: ",array)
 
cocktailSort(array)
print("Array after sorting: ",array)

File: sortingAlgo/cocktailSort/cocktailSort.c
#include <stdio.h>

void cocktail_sort(int *arr, int length)
{
    int start = 0, end = length - 1, flag = 1, temp;
    while (flag)
    {
        flag = 0;
        for (int i = start; i < end; i++)
        {
            if (arr[i] > arr[i + 1])
            {
                temp = arr[i];
                arr[i] = arr[i + 1];
                arr[i + 1] = temp;
                flag = 1;
            }
        }
        if (!flag)
            break;
        flag = 0;
        end--;
        for (int i = end - 1; i >= start; i--)
        {
            if (arr[i] > arr[i + 1])
            {
                temp = arr[i];
                arr[i] = arr[i + 1];
                arr[i + 1] = temp;
                flag = 1;
            }
        }
        start++;
    }
}

void main()
{
    int arr[] = {54, 74, 98, 154, 98, 32, 20, 13, 35, 40}, size = sizeof(arr) / sizeof(arr[0]);
    printf("array before: ");
    for (int i = 0; i < size; i++)
    {
        printf("%d, ", arr[i]);
    }
    printf("\nperforming cocktail sort...");
    cocktail_sort(arr, size);
    printf("\narray after: ");
    for (int i = 0; i < size; i++)
    {
        printf("%d, ", arr[i]);
    }
    printf("\n");
}